### PR TITLE
`gatekeeper.configuration` and `percent_bucket` access

### PR DESCRIPTION
**Pull request by mlebkowski**
@lbarulski, @nediam, I present you:
## Default gates configuration

In the following format:

```
gate_keeper:
    gates:
      sample_gate:
        default: allow-all
      test_feature:
        attributes:
          bucket: 0.1
```

`gatekeeper.configuration` class is responsible for returning those values for given gate name. The `Voter` is using it to provide default attributes to `Gatekeeper\Gatekeeper::hasAccess()`
## Percentage Bucket access

Grants access to given percentage of random users and stores this information in their session. This Access uses a funny session key to store information, since the `AccessInterface` doesn’t get the information about the gate it votes on. So users are just tagged as "belongs to the 10-th percentile bucket", etc. 

I will change it if you’ll modify the interface.

The current usage is via configuration (if the `DefaultFallbackRepository` is in use):

```
test_feature:
   default: percentage-bucket
   attributes:
      bucket: 0.2
```

Or when invoking the `Voter`:

``` php
$securityContext->isGranted('GATE_test_feature', ['bucket' => '1.0']);
```
## Gates descriptions

To be done. This feature introduces gate configuration, so there’s possibility to add description and treat this configuration as a list of all available gates for some kind of GateManager interface.

[Review on Reviewable](https://reviewable.io/reviews/znanylekarz/gatekeeper-bundle/1)

### CODE REVIEW

File: Access/PercentageBucket.php [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+class PercentageBucket extends AbstractAccess
+{
+	const KEY_ACCESS = 'percentage-bucket-access-%d';
+
+	/**
+	 * @var Session
+	 */
+	private $session;
+
+	function __construct(Session $session)
```
Comment (lbarulski) [inline Access/PercentageBucket.php]: public

File: DependencyInjection/GateKeeperExtension.php [hunk 2]
```diff
@@ -24,6 +24,7 @@ public function load(array $configs, ContainerBuilder $container)
 
 		$container->setParameter('gatekeeper.repository.service', $config['repository_service']);
 		$container->setParameter('gatekeeper.provider.service', $config['provider_service']);
+		$container->getDefinition('gatekeeper.configuration')->replaceArgument(0, $config['gates']);
```
Comment (lbarulski) [inline DependencyInjection/GateKeeperExtension.php]: bad idea, set parameter and inject them to the service

File: Access/PercentageBucket.php [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+	public function hasAccess()
+	{
+		$requiredBucket = isset($this->attributes['bucket']) ? (float) $this->attributes['bucket'] : 0;
+
+		if (0 >= $requiredBucket)
+		{
+			return false;
+		}
+
+		if (false === $this->session->isStarted())
```
Comment (lbarulski) [inline Access/PercentageBucket.php:43]: `|| true === $request->hasPreviousSession()`
http://api.symfony.com/2.0/Symfony/Component/HttpFoundation/Request.html#method_hasPreviousSession

File: Access/PercentageBucket.php [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+		{
+			return false;
+		}
+
+		$keyName = sprintf(self::KEY_ACCESS, $requiredBucket * 100);
```
Comment (lbarulski) [inline Access/PercentageBucket.php:48]: gate name - shouldn't be used there?

**Comment by lbarulski** [comment]
- [x] tests (now is there a lot of logic)
- [x] docblocks - everywhere
- [x] fix configuration
- [ ] gate name in session key (?)

File: Repository/DynamicGateModel.php [hunk 5]
```diff
@@ -0,0 +1,34 @@
+<?php
+
+namespace GateKeeperBundle\Provider;
+
+use GateKeeper\Model\ModelInterface;
+
+class DynamicGateModel implements ModelInterface
+{
+	private $access;
+	private $gate;
```
Comment (prgTW) [inline Repository/DynamicGateModel.php:16]: docblocks?

File: Service/GateKeeperConfiguration.php [hunk 6]
```diff
@@ -0,0 +1,27 @@
+<?php
+
+namespace GateKeeperBundle\Service;
+
+class GateKeeperConfiguration
+{
+	private $configuration = [];
+
+	public function __construct($configuration)
```
Comment (prgTW) [inline Service/GateKeeperConfiguration.php:9]: `array $...`?

File: Access/PercentageBucket.php (hunk 4 above)
Comment (mlebkowski) [inline Access/PercentageBucket.php:48]: It would be either ugly or a breaking change

File: Access/PercentageBucket.php (hunk 3 above)
Comment (mlebkowski) [inline Access/PercentageBucket.php:43]: Do I really need to depend on the request?

File: Repository/DynamicGateModel.php (hunk 5 above)
Comment (mlebkowski) [inline Repository/DynamicGateModel.php:16]: What on earth would those docblocks add to this class? This entire files fits on my mobile screen, there is nothing that needs further explaining

File: Access/PercentageBucket.php (hunk 3 above)
Comment (lbarulski) [inline Access/PercentageBucket.php:43]: you want to depend on what? forecaster?

File: Access/PercentageBucket.php (hunk 4 above)
Comment (lbarulski) [inline Access/PercentageBucket.php:48]: bc.

File: Repository/DynamicGateModel.php [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+
+	function __construct($gate, $access)
```
Comment (nediam) [inline Repository/DynamicGateModel.php]: public again

File: Repository/DynamicGateModel.php (hunk 5 above)
Comment (nediam) [inline Repository/DynamicGateModel.php:16]: +1 for @prgTW, here and in constructor

Comment (prgTW) [inline Repository/DynamicGateModel.php:16]: readability. Until php supports type hintinh for scalar types I will force to write docblocks

**Comment by mlebkowski** [comment]
@lbarulski, tests and other stuff added

**Comment by patrykwozinski** [comment]
@mlebkowski merge? :>

**Comment by mlebkowski** [comment]
@patrykwozinski, I dont think its ready yet. Let it rest for 5 more years

**Comment by radek-baczynski** [comment]
kto w 2020? dajcie łapki

**Comment by mlebkowski** [comment]
If you wait a bit until monolith runs on 7.4 I can add those damn scalar typehints and we can merge it!

**Pull request closed without merging**